File: Sources/Olaf/Network/OlafMockResponse.swift
import Foundation

/// A fake response returned for matching requests **without hitting the network** (response mocking).
///
/// Lets you test edge cases without touching the real backend: error bodies, empty lists,
/// 5xx scenarios, slow responses (`delaySeconds`), or transport errors (`transportError` — e.g.
/// no internet). Mocked requests are logged normally under the `.network` category and
/// marked as "Mock" in the detail view.
///
/// ```swift
/// OlafNetwork.addMock(OlafMockResponse(
///     urlContains: "/v1/accounts",
///     json: #"{"accounts": []}"#
/// ))
/// OlafNetwork.addMock(.failure(urlContains: "/v1/transfer", error: .timedOut, delaySeconds: 3))
/// ```
///
/// Matching: the URL (lowercase) contains the `urlContains` part and `method` matches
/// (nil = all methods). If multiple mocks match, the **first one added** wins.
/// Capture filters (`includedURLs`/`excludedURLs`) don't affect mocks.
public struct OlafMockResponse: Sendable, Identifiable {

    /// Record identifier (for removing a single mock from the viewer's mock list).
    public let id: UUID

    /// The part the URL must contain (compared lowercase).
    public var urlContains: String
    /// The HTTP method to match (`nil` = all). Compared uppercase.
    public var method: String?
    public var statusCode: Int
    public var headers: [String: String]
    public var body: Data
    /// The response is delayed by this many seconds (slow network simulation; shows up in the pending requests bar).
    public var delaySeconds: TimeInterval
    /// If set, returns a **transport error** instead of an HTTP response (e.g. `.notConnectedToInternet`).
    public var transportError: URLError.Code?

    public init(
        urlContains: String,
        method: String? = nil,
        statusCode: Int = 200,
        headers: [String: String] = ["Content-Type": "application/json"],
        body: Data = Data(),
        delaySeconds: TimeInterval = 0,
        transportError: URLError.Code? = nil
    ) {
        self.id = UUID()
        self.urlContains = urlContains.lowercased()
        self.method = method?.uppercased()
        self.statusCode = statusCode
        self.headers = headers
        self.body = body
        self.delaySeconds = max(0, delaySeconds)
        self.transportError = transportError
    }

    /// Shortcut for a mock with a JSON body (`Content-Type: application/json`).
    public init(
        urlContains: String,
        method: String? = nil,
        statusCode: Int = 200,
        json: String,
        delaySeconds: TimeInterval = 0
    ) {
        self.init(
            urlContains: urlContains,
            method: method,
            statusCode: statusCode,
            headers: ["Content-Type": "application/json"],
            body: Data(json.utf8),
            delaySeconds: delaySeconds
        )
    }

    /// Shortcut for a transport-error mock (no response; throws a URLError).
    public static func failure(
        urlContains: String,
        method: String? = nil,
        error: URLError.Code = .notConnectedToInternet,
        delaySeconds: TimeInterval = 0
    ) -> OlafMockResponse {
        OlafMockResponse(
            urlContains: urlContains,
            method: method,
            delaySeconds: delaySeconds,
            transportError: error
        )
    }

    /// Does this mock match the given request?
    func matches(_ request: URLRequest) -> Bool {
        guard let url = request.url?.absoluteString.lowercased(), url.contains(urlContains) else {
            return false
        }
        guard let method else { return true }
        return method == (request.httpMethod ?? "GET").uppercased()
    }
}

File: Android/olaf/src/main/java/com/olaf/network/OlafMockResponse.kt
package com.olaf.network

import okhttp3.MediaType.Companion.toMediaTypeOrNull
import okhttp3.Protocol
import okhttp3.Request
import okhttp3.Response
import okhttp3.ResponseBody.Companion.toResponseBody
import java.io.IOException
import java.net.SocketTimeoutException
import java.net.UnknownHostException
import java.util.UUID

/**
 * A fake response returned for matching requests **without hitting the network**.
 *
 * Lets you exercise edge cases without touching the real backend: error bodies, empty lists, 5xx
 * scenarios, slow responses ([delayMillis]) or transport failures ([transportError] — e.g. no
 * connectivity). Mocked calls are still logged normally and flagged as "Mock" in the detail view.
 *
 * ```kotlin
 * OlafNetwork.addMock(OlafMockResponse(urlContains = "/v1/accounts", json = """{"accounts": []}"""))
 * OlafNetwork.addMock(OlafMockResponse.failure("/v1/rates", TransportError.Timeout, delayMillis = 3_000))
 * ```
 *
 * Matching: the lowercased URL contains [urlContains] and [method] matches (`null` = any method).
 * When several mocks match, the **first one added** wins. Capture filters don't affect mocks.
 */
data class OlafMockResponse(
    /** Fragment the URL must contain; compared lowercase. */
    val urlContains: String,

    /** HTTP method to match (`null` = all). Compared uppercase. */
    val method: String? = null,

    val statusCode: Int = 200,

    val headers: Map<String, String> = mapOf("Content-Type" to "application/json"),

    val body: ByteArray = ByteArray(0),

    /** Delays the response by this many milliseconds — a slow-network simulation. */
    val delayMillis: Long = 0,

    /** When set, a **transport failure** is thrown instead of returning an HTTP response. */
    val transportError: TransportError? = null,

    /** Identifier used to remove a single mock from the viewer's mock list. */
    val id: String = UUID.randomUUID().toString()
) {

    /** Transport-level failures a mock can simulate. */
    enum class TransportError {
        NotConnectedToInternet,
        Timeout,
        HostNotFound;

        internal val message: String
            get() = when (this) {
                NotConnectedToInternet -> "Not connected to the internet"
                Timeout -> "The request timed out"
                HostNotFound -> "Host could not be resolved"
            }

        internal fun toIOException(): IOException = when (this) {
            NotConnectedToInternet -> IOException(message)
            Timeout -> SocketTimeoutException(message)
            HostNotFound -> UnknownHostException(message)
        }
    }

    /** Convenience constructor for a JSON-bodied mock. */
    constructor(
        urlContains: String,
        json: String,
        method: String? = null,
        statusCode: Int = 200,
        delayMillis: Long = 0
    ) : this(
        urlContains = urlContains,
        method = method,
        statusCode = statusCode,
        headers = mapOf("Content-Type" to "application/json"),
        body = json.toByteArray(),
        delayMillis = delayMillis
    )

    /** Does this mock match the given request? */
    internal fun matches(request: Request): Boolean {
        if (!request.url.toString().lowercase().contains(urlContains.lowercase())) return false
        val method = method ?: return true
        return method.uppercase() == request.method.uppercase()
    }

    internal fun toResponse(request: Request): Response {
        val contentType = headers.entries
            .firstOrNull { it.key.equals("Content-Type", ignoreCase = true) }
            ?.value
            ?.toMediaTypeOrNull()

        val builder = Response.Builder()
            .request(request)
            .protocol(Protocol.HTTP_1_1)
            .code(statusCode)
            .message(statusMessage(statusCode))
            .body(body.toResponseBody(contentType))

        headers.forEach { (name, value) -> builder.header(name, value) }
        return builder.build()
    }

    // `body` is a ByteArray, so the generated data-class equality would compare references.
    override fun equals(other: Any?): Boolean = this === other || (other is OlafMockResponse && id == other.id)

    override fun hashCode(): Int = id.hashCode()

    companion object {
        /** Shortcut for a transport-failure mock (no response; an IOException is thrown). */
        fun failure(
            urlContains: String,
            error: TransportError = TransportError.NotConnectedToInternet,
            method: String? = null,
            delayMillis: Long = 0
        ): OlafMockResponse = OlafMockResponse(
            urlContains = urlContains,
            method = method,
            delayMillis = delayMillis,
            transportError = error
        )

        private fun statusMessage(code: Int): String = when (code) {
            200 -> "OK"
            201 -> "Created"
            204 -> "No Content"
            400 -> "Bad Request"
            401 -> "Unauthorized"
            403 -> "Forbidden"
            404 -> "Not Found"
            500 -> "Internal Server Error"
            503 -> "Service Unavailable"
            else -> "Mock"
        }
    }
}
